File: completion_list/210802/241_Different_Ways_to_Add_Parentheses.swift

class Solution {
    private enum CalculationError: Error {
        case notDefinedOperation
    }
    
    private enum Operator: Character {
        case multiply   = "*"
        case plus       = "+"
        case subtract   = "-"
        
        func calculate<T: Numeric>(lhs: T, rhs: T) -> T {
            switch self {
            case .multiply: return lhs * rhs
            case .plus:     return lhs + rhs
            case .subtract: return lhs - rhs
            }
            
        }
    }
    
    private var visited: [String: [Int]]
    
    init() {
        visited = [:]
    }
    
    func diffWaysToCompute(_ expression: String) -> [Int] {
        let results = dfs(expression)
        return results
    }
    
    private func dfs(_ str: String) -> [Int] {
        if let results = visited[str] {
            return results
        }
        
        if let num = Int(str) {
            return [num]
        }
        
        var results: [Int] = []
        
        str.enumerated().forEach { item in
            let char: Character = item.element
            let index: Int = item.offset
            
            guard let oper = Operator(rawValue: char) else {
                return
            }
            
            let leftList = dfs(str[0..<index])
            let rightList = dfs(str[index+1..<str.count])
            
            leftList.forEach { left in
                rightList.forEach { right in
                    results.append(oper.calculate(lhs: left, rhs: right))
                }
            }
        }
        
        visited[str] = results
        return results
    }
}


extension String {
    subscript(bounds: CountableRange<Int>) -> Self {
        let start = index(startIndex, offsetBy: bounds.lowerBound)
        let end = index(startIndex, offsetBy: bounds.upperBound)
        return String(self[start..<end])
    }
}

File: completion_list/210802/241_Different_Ways_to_Add_Parentheses.py

class Solution:
  def __init__(self):
    self.operation = {}
    self.operation["*"] = lambda a, b: a*b
    self.operation["+"] = lambda a, b: a+b
    self.operation["-"] = lambda a, b: a-b
    self.visited = {}

  def calculate(self, lhs, rhs, op):
    return self.operation[op](lhs, rhs)

  def dfs(self, str):
    if str in self.visited:
      return self.visited[str]

    if str.isdigit():
        return [int(str)]
    
    results = []
    for i in range(len(str)):
        if str[i] not in self.operation:
            continue
        op = str[i]
        left = self.dfs(str[:i]) # 모든 연산 케이스
        right = self.dfs(str[i+1:]) 
        
        for l in left:
            for r in right:
                results.append(self.calculate(l,r,op))

    self.visited[str] = results
    return results

  def diffWaysToCompute(self, expression: str):
    self.visited = {}
    results = self.dfs(expression)
    return results



s = Solution() 
print(s.diffWaysToCompute("2*3-4*5"))
print(s.diffWaysToCompute("2-1-1"))

'''
틀렸던 이유

left, right의 한 case만 고려했었음.
모든 case를 고려했어야함.

'''

